feat: update prometheus agent daemonset proposal (#7571)
prometheus-operator/prometheus-operator: Documentation/proposals/202405-agent-daemonset.md

* Allow users to set the priority of Prometheus Agent pod compared to other pods on the same node, if desired.
* Allow each Prometheus Agent pod to only scrape from the pods from PodMonitor that run on the same node.

### 4.1. ServiceMonitor Support with EndpointSlice (Extended Goal)

Building upon the MVP, enable ServiceMonitor support for DaemonSet mode by leveraging EndpointSlice discovery to address the performance concerns mentioned in the non-goals section. This extended implementation includes:

* **EndpointSlice Integration**: Implement support for `serviceDiscoveryRole: EndpointSlice` in PrometheusAgent CRD to use EndpointSlice API instead of classic Endpoints for service discovery.
* **Performance Optimization**: Reduce Kubernetes API server load by using the more efficient and scalable EndpointSlice API, which distributes endpoint information across multiple smaller objects instead of one large Endpoints object.
* **DaemonSet Compatibility**: Ensure EndpointSlice discovery works correctly with DaemonSet mode for node-local service monitoring.

> * **Validation and Fallback**: Implement proper validation to ensure EndpointSlice is only used when supported by the Kubernetes cluster (v1.21+), with graceful fallback to classic Endpoints on older clusters.

Reviewer: I am wondering whether we need to provide fallback as 1.21 itself is way too older version
@simonpasquier wdyt?

* Shard
* Storage

We will add a new `mode` field that accepts either `StatefulSet` or `DaemonSet`, with `StatefulSet` being the default. If the DaemonSet mode is activated (`mode: DaemonSet`), all the unrelated fields listed above will not be accepted. In the MVP, we will simply fail the reconciliation if any of those fields are set. We will prevent users to directly switch from a live StatefulSet setup to DaemonSet, because that might break their workload if they forget to unset the unsupported fields.

#### 6.1.1 CEL Validation rules

To prevent users from accidentally providing invalid configurations, we will implement validation rules using [Kubernetes' Common Expression Language (CEL)](https://kubernetes.io/docs/reference/using-api/cel/).

> These CEL rules will affect the DaemonSet mode and prevent:

Reviewer 2: (nit)
```suggestion
When `mode: Daemonset`, the following CEL rules will be applied:

Disallowed Configuration Options:
```

Author: https://github.com/prometheus-operator/prometheus-operator/blob/9be68efafca772687346ffa8b8cc0cbeba88d1e7/pkg/apis/monitoring/v1alpha1/prometheusagent_types.go#L111

Done

> - Setting replicas 

Reviewer 2: ```suggestion
- `replicas`
```

- Setting storage configuration 

> - Using ServiceMonitorSelector 

Reviewer 2: ```suggestion
- `serviceMonitorSelector` 
```

> - Setting shards 

Reviewer 2: ```suggestion
- `shards`
```

- Setting persistentVolumeClaimRetentionPolicy 

> - Using serviceMonitorNamespaceSelector 

Reviewer 2: ```suggestion
- `serviceMonitorNamespaceSelector` 
```

> - Setting serviceName 

Reviewer 2: hmm... I am not sure if this is unacceptable for a Daemonset. I think it is a valid use case to have a governing service that exposes all the Daemonsets.
CC: @simonpasquier @slashpai

Additionally, it enforces the following rules:

> - Require PodMonitorSelector in DaemonSet mode

Author: Yup looked into this and it's optional so I've removed it now.

- Using VolumeClaimTemplates in DaemonSet mode

This is implemented by adding `x-kubernetes-validations` like:

```yaml
x-kubernetes-validations:
  - rule: "self.mode == 'DaemonSet' ? !has(self.replicas) : true"
    message: "replicas field is not allowed when mode is 'DaemonSet'"
```

#### 6.1.2 Runtime Validation Logic as Fallback 

CEL validation will provide immediate feedback during `kubectl apply` but we will need runtime validation logic in the controller as a fallback mechanism. This fallback will be integrated directly in the `PrometheusAgent` reconciler loop.

This is mainly because :

> 1. CEL validation will require Kubernetes version 1.25+ and hence not all users might have CEL supported clusters.

Reviewer 2: This is a valid concern. But Kubernetes is currently at major release `v1.33.0`. Would it be fine to not support this as well? Thoughts? @simonpasquier @slashpai @nicolastakashi
